packages/ethereum/contracts/ERC777/ERC777Snapshot.sol

```diff
@@ -14,171 +11,147 @@ import "@openzeppelin/contracts/token/ERC777/ERC777.sol";
-        //  requested id, and its value is the one to return.
-        //  c) More snapshots were created after the requested one, and the queried value was later modified. There will be
-        //  no entry for the requested id: the value that corresponds to it is that of the smallest snapshot id that is
-        //  larger than the requested one.
-        //
-        // In summary, we need to find an element in an array, returning the index of the smallest value that is larger if
-        // it is not found, unless said value doesn't exist (e.g. when all values are smaller). Arrays.findUpperBound does
-        // exactly this.
-
-        uint256 index = snapshots.ids.findUpperBound(snapshotId);
-
-        if (index == snapshots.ids.length) {
-            return (false, 0);
-        } else {
-            return (true, snapshots.values[index]);
+        super._beforeTokenTransfer(operator, from, to, amount);
+
+        if (from == address(0)) {
+            // mint
+            updateValueAtNow(accountSnapshots[to], balanceOf(to).add(amount));
+            updateValueAtNow(totalSupplySnapshots, totalSupply().add(amount));
+        } else if (to == address(0)) {
+            // burn
+            updateValueAtNow(accountSnapshots[from], balanceOf(from).sub(amount));
+            updateValueAtNow(totalSupplySnapshots, totalSupply().sub(amount));
+        } else if (from != to) {
+            // transfer
+            updateValueAtNow(accountSnapshots[from], balanceOf(from).sub(amount));
+            updateValueAtNow(accountSnapshots[to], balanceOf(to).add(amount));
         }
     }
 
-    function _updateAccountSnapshot(address account) private {
-        _updateSnapshot(_accountBalanceSnapshots[account], balanceOf(account));
-    }
-
-    function _updateTotalSupplySnapshot() private {
-        _updateSnapshot(_totalSupplySnapshots, totalSupply());
-    }
+    /**
+     * @dev `_valueAt` retrieves the number of tokens at a given block number
+     * @param snapshots The history of values being queried
+     * @param _block The block number to retrieve the value at
+     * @return The number of tokens being queried
+     */
+    function _valueAt(
+        Snapshot[] storage snapshots,
+        uint _block
+    ) view internal returns (uint) {
+        if (snapshots.length == 0) return 0;
+
+        // Shortcut for the actual value
+        if (_block >= snapshots[snapshots.length - 1].fromBlock) {
+            return snapshots[snapshots.length - 1].value;
+        }
+        if (_block < snapshots[0].fromBlock) {
+            return 0;
+        }
 
-    function _updateSnapshot(Snapshots storage snapshots, uint256 currentValue) private {
-        uint256 currentId = _currentSnapshotId.current();
-        if (_lastSnapshotId(snapshots.ids) < currentId) {
-            snapshots.ids.push(currentId);
-            snapshots.values.push(currentValue);
+        // Binary search of the value in the array
+        uint min = 0;
```

Comment:
we should be consistent about using `uint256` vs alias `uint`, I prefer the former but not 100% what's best practice right now.

Reply:
agree on `uint256` 👍